File: composer/src/main/java/io/piano/android/composer/PrefsStorage.java
package io.piano.android.composer;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

import java.util.Date;

@RestrictTo(RestrictTo.Scope.LIBRARY)
class PrefsStorage {
    private static final String PREFS_NAME = "io.piano.android.composer";
    static final String KEY_VISIT_ID = "visitId";
    static final String KEY_VISIT_ID_TIMESTAMP = "visitIdTimestampMillis";
    static final String KEY_TP_BROWSER_COOKIE = "tbc";
    static final String KEY_XBUILDER_BROWSER_COOKIE = "xbc";
    static final String KEY_TP_ACCESS_COOKIE = "tac";
    static final String KEY_TIMEZONE_OFFSET = "timeZoneOffsetMillis";
    static final String KEY_VISIT_TIMEOUT = "visitTimeoutMinutes";

    private final SharedPreferences prefs;

    PrefsStorage(@NonNull Context context) {
        this.prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }

    @Nullable
    String getValueForKey(@NonNull String key) {
        return prefs.getString(key, null);
    }

    void setValueForKey(@NonNull String key, @Nullable String value) {
        prefs.edit().putString(key, value).apply();
    }

    @Nullable
    String getVisitId() {
        return getValueForKey(KEY_VISIT_ID);
    }

    void setVisitId(@Nullable String value) {
        setValueForKey(KEY_VISIT_ID, value);
    }

    long getVisitTimestamp() {
        return prefs.getLong(KEY_VISIT_ID_TIMESTAMP, 0);
    }

    void setVisitDate(long value) {
        prefs.edit().putLong(KEY_VISIT_ID_TIMESTAMP, value).apply();
    }

    void setVisitDate(@Nullable Date value) {
        setVisitDate(value != null ? value.getTime() : 0);
    }

    @Nullable
    String getTpBrowserCookie() {
        return getValueForKey(KEY_TP_BROWSER_COOKIE);
    }

    void setTpBrowserCookie(@Nullable String value) {
        setValueForKey(KEY_TP_BROWSER_COOKIE, value);
    }

    @Nullable
    String getXbuilderBrowserCookie() {
        return getValueForKey(KEY_XBUILDER_BROWSER_COOKIE);
    }

    void setXbuilderBrowserCookie(@Nullable String value) {
        setValueForKey(KEY_XBUILDER_BROWSER_COOKIE, value);
    }

    @Nullable
    String getTpAccessCookie() {
        return getValueForKey(KEY_TP_ACCESS_COOKIE);
    }

    void setTpAccessCookie(@Nullable String value) {
        setValueForKey(KEY_TP_ACCESS_COOKIE, value);
    }

    int getServerTimezoneOffset() {
        return prefs.getInt(KEY_TIMEZONE_OFFSET, 0);
    }

    void setServerTimezoneOffset(int milliseconds) {
        prefs.edit().putInt(KEY_TIMEZONE_OFFSET, milliseconds).apply();
    }

    long getVisitTimeout() {
        return prefs.getLong(KEY_VISIT_TIMEOUT, Constants.VISIT_TIMEOUT_FALLBACK);
    }

    void setVisitTimeout(long milliseconds) {
        prefs.edit().putLong(KEY_VISIT_TIMEOUT, milliseconds).apply();
    }
}

File: composer/src/main/java/io/piano/android/composer/PrefsStorage.kt
package io.piano.android.composer

import android.content.Context
import android.content.SharedPreferences
import androidx.annotation.RestrictTo
import java.util.Date
import java.util.concurrent.TimeUnit

@RestrictTo(RestrictTo.Scope.LIBRARY)
internal class PrefsStorage(context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    internal fun clear() {
        prefs.edit().clear().apply()
    }

    internal fun getValueForKey(key: String): String? {
        return prefs.getString(key, null)
    }

    internal fun setValueForKey(key: String, value: String?) {
        prefs.edit().putString(key, value).apply()
    }

    var visitId: String?
        get() = getValueForKey(KEY_VISIT_ID)
        set(value) {
            setValueForKey(KEY_VISIT_ID, value)
        }

    val visitTimestamp: Long
        get() = prefs.getLong(KEY_VISIT_ID_TIMESTAMP, 0)

    fun setVisitDate(value: Long) {
        prefs.edit().putLong(KEY_VISIT_ID_TIMESTAMP, value).apply()
    }

    fun setVisitDate(value: Date?) {
        setVisitDate(value?.time ?: 0)
    }

    var tpBrowserCookie: String
        get() = getValueForKey(KEY_TP_BROWSER_COOKIE).orEmpty()
        set(value) {
            setValueForKey(KEY_TP_BROWSER_COOKIE, value)
        }

    var xbuilderBrowserCookie: String
        get() = getValueForKey(KEY_XBUILDER_BROWSER_COOKIE).orEmpty()
        set(value) {
            setValueForKey(KEY_XBUILDER_BROWSER_COOKIE, value)
        }

    var tpAccessCookie: String
        get() = getValueForKey(KEY_TP_ACCESS_COOKIE).orEmpty()
        set(value) {
            setValueForKey(KEY_TP_ACCESS_COOKIE, value)
        }

    var serverTimezoneOffset: Int
        get() = prefs.getInt(KEY_TIMEZONE_OFFSET, 0)
        set(milliseconds) {
            prefs.edit().putInt(KEY_TIMEZONE_OFFSET, milliseconds).apply()
        }

    var visitTimeout: Long
        get() = prefs.getLong(
            KEY_VISIT_TIMEOUT,
            VISIT_TIMEOUT_FALLBACK
        )
        set(milliseconds) {
            prefs.edit().putLong(KEY_VISIT_TIMEOUT, milliseconds).apply()
        }

    companion object {
        internal const val PREFS_NAME = "io.piano.android.composer"
        internal const val KEY_VISIT_ID = "visitId"
        internal const val KEY_VISIT_ID_TIMESTAMP = "visitIdTimestampMillis"
        internal const val KEY_TP_BROWSER_COOKIE = "tbc"
        internal const val KEY_XBUILDER_BROWSER_COOKIE = "xbc"
        internal const val KEY_TP_ACCESS_COOKIE = "tac"
        internal const val KEY_TIMEZONE_OFFSET = "timeZoneOffsetMillis"
        internal const val KEY_VISIT_TIMEOUT = "visitTimeoutMinutes"
        internal val VISIT_TIMEOUT_FALLBACK = TimeUnit.MILLISECONDS.convert(30, TimeUnit.MINUTES)
    }
}
